src/lib.rs

```diff
@@ -195,6 +195,25 @@ pub trait MysqlShim<W: Read + Write> {
     fn tls_config(&self) -> Option<std::sync::Arc<rustls::ServerConfig>> {
         None
     }
+
+    /// Called after successful authentication (including TLS if applicable) passing relevant
+    /// information to allow additional logic in the MySqlShim implementation.
+    fn after_authentication(
+        &mut self,
+        _context: &AuthenticationContext,
+    ) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+/// Information about an authenticated user
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct AuthenticationContext {
+    /// The username exactly as passed by the client,
+    pub username: Option<Vec<u8>>,
+    #[cfg(feature = "tls")]
+    /// The TLS certificate chain presented by the client.
+    pub tls_client_certs: Option<Vec<rustls::Certificate>>,
```

Comment:
I'd like to expose this as `&[]` instead so we don't have to copy the certs around. I understand that it's tricky for `username` since it could come from the first _or_ second packet, but for the certs I think it should be doable.